synapse/storage/databases/main/events_worker.py

```diff
@@ -901,21 +902,25 @@ async def get_stripped_room_state_from_event_context(
         Returns:
             A list of dictionaries, each representing a stripped state event from the room.
         """
-        current_state_ids = await context.get_current_state_ids()
+        if membership_user_id:
+            types = chain(
+                state_keys_to_include.to_types(),
+                [(EventTypes.Member, membership_user_id)],
+            )
+            filter = StateFilter.from_types(types)
+        else:
+            filter = state_keys_to_include
+        selected_state_ids = await context.get_current_state_ids(filter)
 
         # We know this event is not an outlier, so this must be
         # non-None.
-        assert current_state_ids is not None
-
-        # The state to include
-        state_to_include_ids = [
-            e_id
-            for k, e_id in current_state_ids.items()
-            if k[0] in state_types_to_include
-            or (membership_user_id and k == (EventTypes.Member, membership_user_id))
-        ]
+        assert selected_state_ids is not None
+
+        # Confusingly, get_current_state_events may return events that are discarded by
+        # the filter, if they're in context._state_delta_due_to_event. Strip these away.
```

Comment:
hmph. Sounds footgunny. Might be worth looking to see if it's worth pushing `StateFilter.filter` down into `get_current_state_ids`?
Possibly it may not be (if not every caller does that for some reason).

(N.B. you don't have to — merely a suggestion if it's an easy improvement whilst in the area)

Reply:
I considered it, but I'm pretty scared of the StateFilter machinery. I'll have a quick look at the call sites tomorrow

Reply:
Quick search for call sites that provide a state filter:

https://github.com/matrix-org/synapse/blob/b5b5f6608462a988b05502a3b70b6a57ca3846d2/synapse/handlers/federation_event.py#L1986-L1990

https://github.com/matrix-org/synapse/blob/b5b5f6608462a988b05502a3b70b6a57ca3846d2/synapse/handlers/message.py#L232-L235

https://github.com/matrix-org/synapse/blob/b5b5f6608462a988b05502a3b70b6a57ca3846d2/synapse/handlers/register.py#L541-L543

https://github.com/matrix-org/synapse/blob/b5b5f6608462a988b05502a3b70b6a57ca3846d2/synapse/handlers/room.py#L306

https://github.com/matrix-org/synapse/blob/b5b5f6608462a988b05502a3b70b6a57ca3846d2/synapse/handlers/room.py#L484-L490

https://github.com/matrix-org/synapse/blob/b5b5f6608462a988b05502a3b70b6a57ca3846d2/synapse/handlers/room.py#L562-L567

There are more. I couldn't see anywhere that does anything like this.

But I also noticed this comment:

https://github.com/matrix-org/synapse/blob/b5b5f6608462a988b05502a3b70b6a57ca3846d2/synapse/storage/controllers/state.py#L405-L417

so maybe using a StateFilter is counterproductive if it bypasses caching? (I guess it depends on how likely it is that the room's current state exists in memory, how much effort a filter saves when making the db query etc)

Anyway I think this is one to leave for a rainy day!